kolibri/core/auth/management/commands/bulkimportusers.py

```diff
@@ -0,0 +1,587 @@
+
+    return checker
+
+
+def enumeration(*args):
+    """
+    Return a value check function which raises a ValueError if the value (case
+    insensitive) is not in the enumeration of values provided by args.
+    """
+    assert len(args) > 0, "at least one argument is required"
+    if len(args) == 1:
+        # assume the first argument defines the membership
+        members = args[0].lower()
+    else:
+        members = tuple(map(str.lower, args))
+
+    def checker(value):
+        if value.lower() not in members:
+            raise ValueError(value)
+
+    return checker
+
+
+def valid_name(username=True, null=False):
+    """
+    Return a value check function which raises a ValueError if the value has
+    some of the punctuaction chars that are not allowed.
+    If username is False it allows spaces, slashes and hyphens.
+    If null is not True raises a ValueError if the supplied value is None.
+    """
+
+    def checker(v):
+        if null and v is None:
+            return checker
+        has_punc = "/[\s`~!@#$%^&*()\-+={}\[\]\|\\\/:;\"'<>,\.\?]/"  # noqa
+        if not username:
+            has_punc = "/[`~!@#$%^&*()\+={}\[\]\|\\\/:;\"'<>\.\?]/"  # noqa
+        if re.match(has_punc, v):
+            raise ValueError(v)
+
+    return checker
+
+
+class Validator(object):
+    """
+    Class to apply different validation checks on a CSV data reader.
+    """
+
+    def __init__(self, header_names):
+        self._header_names = header_names
+        self._checks = list()
+        self.classrooms = dict()
+        self.coach_classrooms = dict()
+        self.users = dict()
+        self.roles = {r: list() for r in roles_map.values() if r is not None}
+
+    def add_check(self, header_name, check, message):
+        """
+        Add a header check, i.e., check whether the header record is consistent
+        with the expected field names.
+        `header_name` - name of the header for the column to be checked
+        `check`- function to be used as validator of the values in the column
+        `message` - problem message to report if a value is not valid
+        """
+        self._checks.append((header_name, check, message))
+
+    def get_username(self, row):
+        username = row.get("Username")
+        if username in self.users.keys():
+            return None
+
+        return username
+
+    def check_classroom(self, row, username):
+        def append_users(group, key):
+            try:
+                groups = row.get(key, None).split(",")
+                for classroom in groups:
+                    if not classroom:
+                        continue
+                    if classroom in group:
+                        group[classroom].append(username)
+                    else:
+                        group[classroom] = [
+                            username,
+                        ]
+            except AttributeError:
+                # there are not members of 'key'
+                pass
```

Comment:
Which lines are expected to throw `AttributeError`? Maybe the `try`...`except` can wrap a smaller number of lines?

Also, a `group` in Kolibri is a collection inside a class, so the variable name here is a bit confusing. Maybe `class_list`?

Reply:
for the AttributeErrork, it's line 144. If there is not group with that key, split can't be executed. I will refactor it to apply the try only to that line.

About the name group, yes it makes sense in a kolibri context. I was thinking of it just as its meaning. I  have changed it as you say. I am afraid this kind of naming conflict might happen more times in the code.